pkg/cached/cache.go

```diff
@@ -55,51 +57,65 @@ import (
 )
 
 // Result is the content returned from a call to a cache. It can either
-// be created with [NewResultOK] if the call was a success, or
-// [NewResultErr] if the call resulted in an error.
+// be created with [NewResult] if the call was a success, or
+// [NewError] if the call resulted in an error.
 type Result[T any] struct {
 	Data T
 	Etag string
 	Err  error
 }
 
-// NewResultOK creates a new [Result] for a successful operation.
-func NewResultOK[T any](data T, etag string) Result[T] {
+// NewResult creates a new [Result] for a successful operation.
+func NewResult[T any](data T, etag string) Result[T] {
 	return Result[T]{
 		Data: data,
 		Etag: etag,
 	}
 }
 
-// NewResultErr creates a new [Result] when an error has happened.
-func NewResultErr[T any](err error) Result[T] {
+// NewError creates a new [Result] when an error has happened.
+func NewError[T any](err error) Result[T] {
 	return Result[T]{
 		Err: err,
 	}
 }
 
-// Result can be treated as a [Data] if necessary.
+// Errorf creates a new [Result] when an error has happened.
+func Errorf[T any](s string, params ...any) Result[T] {
+	return NewError[T](fmt.Errorf(s, params...))
+}
+
+// Get returns the cached result. Result can be treated as a [Data] if necessary.
 func (r Result[T]) Get() Result[T] {
 	return r
 }
 
+// Unpack return the individual elements of the result, for more
+// convenient usage:
+//
+//	if value, _, err := c.Get().Unpack(); err != nil {
+//	  ...
+//	}
+func (r Result[T]) Unpack() (T, string, error) {
+	return r.Data, r.Etag, r.Err
+}
+
 // Data is a cache that performs an action whose result data will be
 // cached. It also returns an "etag" identifier to version the cache, so
 // that the caller can know if they have the most recent version of the
 // cache (and can decide to cache some operation based on that).
 //
-// The [NewMerger] and [NewTransformer] automatically handle
+// The [Merge] and [Transform] automatically handle
 // that for you by checking if the etag is updated before calling the
 // merging or transforming function.
 type Data[T any] interface {
-	// Returns the cached data, as well as an "etag" to identify the
+	// Get returns the cached data, as well as an "etag" to identify the
 	// version of the cache, or an error if something happened.
 	Get() Result[T]
 }
 
-// NewMerger creates a new merge cache, a cache that merges the result
-// of other caches. The function only gets called if any of the
-// dependency has changed.
+// Merge merges the result of other caches. The function only gets
```

Comment:
Question: Is it convention to remove the `New` for constructors? I feel like it's easy to get `Merge` confused with a function that merges vs returning a cache that merges on a `Get()`

Reply:
These are logically not constructors, but verbs. That's the whole point of the PR. Promise-like code that mocks a non-promise-like interface. It's a detail that behind the scenes constructors create a datastructure.